File: modules/agent/harness_handoff.go
package agent

import (
	"context"
	"fmt"
	"sync"

	"github.com/woyin/OrangeCoding/modules/ai"
	"github.com/woyin/OrangeCoding/modules/core"
)

// HandoffRequest transfers control from one agent to another.
type HandoffRequest struct {
	FromAgentID   core.AgentId
	ToAgentID     core.AgentId
	Task          string
	Conversation  []core.Message
	ToolCallsMade uint32
	MemoryKeys    []string
	Metadata      map[string]string
}

// HandoffResult is returned when an agent completes a handoff.
type HandoffResult struct {
	FromAgentID   core.AgentId
	ToAgentID     core.AgentId
	Completed     bool
	ToolCallsMade uint32
	Summary       string
	Error         string
}

// HandoffHandler processes agent handoffs.
type HandoffHandler interface {
	CanHandoff(ctx context.Context, req HandoffRequest) (bool, error)
	ExecuteHandoff(ctx context.Context, req HandoffRequest) (HandoffResult, error)
}

// ToolUseBudget tracks per-tool call counts.
type ToolUseBudget struct {
	// MaxUses is the maximum allowed calls per tool. 0 = unlimited.
	MaxUses map[string]uint32
	// counts tracks actual usage.
	counts map[string]uint32
}

// NewToolUseBudget creates a tool use budget.
func NewToolUseBudget() *ToolUseBudget {
	return &ToolUseBudget{
		MaxUses: make(map[string]uint32),
		counts:  make(map[string]uint32),
	}
}

// SetMaxUses sets the maximum allowed calls for a tool.
func (b *ToolUseBudget) SetMaxUses(toolName string, max uint32) {
	b.MaxUses[toolName] = max
}

// RecordCall records that a tool was called. Returns true if the call is allowed.
func (b *ToolUseBudget) RecordCall(toolName string) bool {
	max, hasMax := b.MaxUses[toolName]
	if !hasMax || max == 0 {
		b.counts[toolName]++
		return true
	}
	current := b.counts[toolName]
	if current >= max {
		return false
	}
	b.counts[toolName]++
	return true
}

// Remaining returns how many more calls are allowed for a tool.
func (b *ToolUseBudget) Remaining(toolName string) uint32 {
	max, hasMax := b.MaxUses[toolName]
	if !hasMax || max == 0 {
		return ^uint32(0) // unlimited
	}
	used := b.counts[toolName]
	if used >= max {
		return 0
	}
	return max - used
}

// Used returns how many times a tool has been called.
func (b *ToolUseBudget) Used(toolName string) uint32 {
	return b.counts[toolName]
}

// AgentModelSettings holds per-agent model configuration.
type AgentModelSettings struct {
	Model           string
	Temperature     *float64
	TopP            *float64
	MaxTokens       *uint32
	ReasoningEffort ReasoningEffort
	ReasoningBudget *uint32
}

// ApplyToChatOptions applies agent-specific model settings to ChatOptions.
// Non-zero/non-nil values override the existing options.
func (s AgentModelSettings) ApplyToChatOptions(opts ai.ChatOptions) ai.ChatOptions {
	if s.Model != "" {
		opts.Model = s.Model
	}
	if s.Temperature != nil {
		opts.Temperature = s.Temperature
	}
	if s.TopP != nil {
		opts.TopP = s.TopP
	}
	if s.MaxTokens != nil {
		opts.MaxTokens = s.MaxTokens
	}
	if s.ReasoningEffort != "" {
		opts.ReasoningEffort = string(s.ReasoningEffort)
	}
	if s.ReasoningBudget != nil {
		opts.ReasoningBudgetTokens = s.ReasoningBudget
	}
	return opts
}

// OrchestratorTask represents a decomposed task for an agent.
type OrchestratorTask struct {
	ID          string
	AgentID     core.AgentId
	Description string
	Scope       []string // file paths or directories
	DependsOn   []string // task IDs this depends on
	Priority    int
}

// OrchestratorResult summarizes the outcome of a decomposed task.
type OrchestratorResult struct {
	TaskID  string
	Success bool
	Summary string
	Error   string
}

// Orchestrator decomposes a task and coordinates sub-agents.
type Orchestrator struct {
	mu      sync.RWMutex
	tasks   map[string]OrchestratorTask
	results map[string]OrchestratorResult
}

// NewOrchestrator creates a new orchestrator.
func NewOrchestrator() *Orchestrator {
	return &Orchestrator{
		tasks:   make(map[string]OrchestratorTask),
		results: make(map[string]OrchestratorResult),
	}
}

// AddTask registers a task for orchestration.
func (o *Orchestrator) AddTask(task OrchestratorTask) error {
	if task.ID == "" {
		return fmt.Errorf("orchestrator: task ID is required")
	}
	o.mu.Lock()
	defer o.mu.Unlock()
	o.tasks[task.ID] = task
	return nil
}

// RecordResult records the outcome of a task.
func (o *Orchestrator) RecordResult(result OrchestratorResult) {
	o.mu.Lock()
	defer o.mu.Unlock()
	o.results[result.TaskID] = result
}

// ReadyTasks returns tasks whose dependencies are all completed successfully.
func (o *Orchestrator) ReadyTasks() []OrchestratorTask {
	o.mu.RLock()
	defer o.mu.RUnlock()
	var ready []OrchestratorTask
	for _, task := range o.tasks {
		if _, done := o.results[task.ID]; done {
			continue
		}
		allDepsMet := true
		for _, depID := range task.DependsOn {
			result, exists := o.results[depID]
			if !exists || !result.Success {
				allDepsMet = false
				break
			}
		}
		if allDepsMet {
			ready = append(ready, task)
		}
	}
	return ready
}

// AllCompleted returns true if all tasks have results.
func (o *Orchestrator) AllCompleted() bool {
	o.mu.RLock()
	defer o.mu.RUnlock()
	for _, task := range o.tasks {
		if _, exists := o.results[task.ID]; !exists {
			return false
		}
	}
	return true
}

// Summary returns a summary of all task results.
func (o *Orchestrator) Summary() string {
	o.mu.RLock()
	defer o.mu.RUnlock()
	total := len(o.tasks)
	completed := 0
	failed := 0
	for _, r := range o.results {
		if r.Success {
			completed++
		} else {
			failed++
		}
	}
	return fmt.Sprintf("tasks: %d total, %d completed, %d failed", total, completed, failed)
}

File: packages/agent/src/harness-handoff.ts
/**
 * @module harness-handoff
 *
 * Agent handoff protocol for transitioning between agent instances.
 *
 * When an agent session needs to be transferred to a different agent
 * (e.g., model upgrade, specialized agent, or session resumption),
 * the handoff module manages:
 * - State serialization and transfer
 * - Context preservation
 * - Conversation history migration
 * - Graceful transition with minimal disruption
 */

import type { AgentId, Message } from "@orangecoding/core";
import type { ChatOptions } from "@orangecoding/ai";
import type { ReasoningEffort } from "./harness-profile.js";

// ---------------------------------------------------------------------------
// HandoffRequest
// ---------------------------------------------------------------------------

export interface HandoffRequest {
  fromAgentID: AgentId;
  toAgentID: AgentId;
  task: string;
  conversation: Message[];
  toolCallsMade: number;
  memoryKeys: string[];
  metadata: Record<string, string>;
}

// ---------------------------------------------------------------------------
// HandoffResult
// ---------------------------------------------------------------------------

export interface HandoffResult {
  fromAgentID: AgentId;
  toAgentID: AgentId;
  completed: boolean;
  toolCallsMade: number;
  summary: string;
  error: string;
}

// ---------------------------------------------------------------------------
// HandoffHandler
// ---------------------------------------------------------------------------

export interface HandoffHandler {
  canHandoff(signal: AbortSignal | undefined, req: HandoffRequest): Promise<[boolean, Error | null]>;
  executeHandoff(signal: AbortSignal | undefined, req: HandoffRequest): Promise<[HandoffResult, Error | null]>;
}

// ---------------------------------------------------------------------------
// ToolUseBudget
// ---------------------------------------------------------------------------

export class ToolUseBudget {
  maxUses: Map<string, number>;
  private _counts: Map<string, number>;

  constructor() {
    this.maxUses = new Map();
    this._counts = new Map();
  }

  /** SetMaxUses sets the maximum allowed calls for a tool. */
  setMaxUses(toolName: string, max: number): void {
    this.maxUses.set(toolName, max);
  }

  /** RecordCall records that a tool was called. Returns true if the call is allowed. */
  recordCall(toolName: string): boolean {
    const max = this.maxUses.get(toolName);
    if (max === undefined || max === 0) {
      this._counts.set(toolName, (this._counts.get(toolName) ?? 0) + 1);
      return true;
    }
    const current = this._counts.get(toolName) ?? 0;
    if (current >= max) return false;
    this._counts.set(toolName, current + 1);
    return true;
  }

  /** Remaining returns how many more calls are allowed for a tool. */
  remaining(toolName: string): number {
    const max = this.maxUses.get(toolName);
    if (max === undefined || max === 0) return Number.MAX_SAFE_INTEGER;
    const used = this._counts.get(toolName) ?? 0;
    if (used >= max) return 0;
    return max - used;
  }

  /** Used returns how many times a tool has been called. */
  used(toolName: string): number {
    return this._counts.get(toolName) ?? 0;
  }
}

// ---------------------------------------------------------------------------
// AgentModelSettings
// ---------------------------------------------------------------------------

export interface AgentModelSettings {
  model?: string;
  temperature?: number;
  topP?: number;
  maxTokens?: number;
  reasoningEffort?: ReasoningEffort;
  reasoningBudget?: number;
}

/** ApplyToChatOptions applies agent-specific model settings to ChatOptions. */
export function applyModelSettingsToChatOptions(settings: AgentModelSettings, opts: ChatOptions): ChatOptions {
  if (settings.model) opts.model = settings.model;
  if (settings.temperature !== undefined) opts.temperature = settings.temperature;
  if (settings.topP !== undefined) opts.top_p = settings.topP;
  if (settings.maxTokens !== undefined) opts.max_tokens = settings.maxTokens;
  if (settings.reasoningEffort) opts.reasoning_effort = settings.reasoningEffort;
  if (settings.reasoningBudget !== undefined) opts.reasoning_budget_tokens = settings.reasoningBudget;
  return opts;
}

// ---------------------------------------------------------------------------
// OrchestratorTask
// ---------------------------------------------------------------------------

export interface OrchestratorTask {
  id: string;
  agentID: AgentId;
  description: string;
  scope: string[]; // file paths or directories
  dependsOn: string[]; // task IDs this depends on
  priority: number;
}

// ---------------------------------------------------------------------------
// OrchestratorResult
// ---------------------------------------------------------------------------

export interface OrchestratorResult {
  taskID: string;
  success: boolean;
  summary: string;
  error: string;
}

// ---------------------------------------------------------------------------
// Orchestrator
// ---------------------------------------------------------------------------

export class Orchestrator {
  private _tasks: Map<string, OrchestratorTask>;
  private _results: Map<string, OrchestratorResult>;

  constructor() {
    this._tasks = new Map();
    this._results = new Map();
  }

  /** AddTask registers a task for orchestration. */
  addTask(task: OrchestratorTask): void {
    if (!task.id) throw new Error("orchestrator: task ID is required");
    this._tasks.set(task.id, task);
  }

  /** RecordResult records the outcome of a task. */
  recordResult(result: OrchestratorResult): void {
    this._results.set(result.taskID, result);
  }

  /** ReadyTasks returns tasks whose dependencies are all completed successfully. */
  readyTasks(): OrchestratorTask[] {
    const ready: OrchestratorTask[] = [];
    for (const task of this._tasks.values()) {
      if (this._results.has(task.id)) continue;
      let allDepsMet = true;
      for (const depID of task.dependsOn) {
        const result = this._results.get(depID);
        if (!result || !result.success) {
          allDepsMet = false;
          break;
        }
      }
      if (allDepsMet) ready.push(task);
    }
    return ready;
  }

  /** AllCompleted returns true if all tasks have results. */
  get allCompleted(): boolean {
    for (const task of this._tasks.keys()) {
      if (!this._results.has(task)) return false;
    }
    return true;
  }

  /** Summary returns a summary of all task results. */
  summary(): string {
    const total = this._tasks.size;
    let completed = 0;
    let failed = 0;
    for (const r of this._results.values()) {
      if (r.success) completed++;
      else failed++;
    }
    return `tasks: ${total} total, ${completed} completed, ${failed} failed`;
  }
}
